ariadne/src/Ariadne/Wallet/UiAdapter.hs

```diff
@@ -0,0 +1,133 @@
+module Ariadne.Wallet.UiAdapter
+  ( UiWalletData(..)
+  , UiAccountData(..)
+  , UiWalletSelection(..)
+  -- * Lens
+  , uwdAccounts
+  , uadAddresses
+  --
+  , toUiWalletDatas
+  , toUiWalletSelection
+  ) where
+
+import Universum
+
+import qualified Data.Vector as V
+
+import Ariadne.Wallet.Cardano.Kernel.DB.AcidState
+import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet
+import Ariadne.Wallet.Cardano.Kernel.DB.HdWallet.Read
+import Ariadne.Wallet.Cardano.Kernel.DB.InDb
+import Ariadne.Wallet.Cardano.Kernel.DB.Util.IxSet
+import Ariadne.Wallet.Face
+
+import Control.Lens (makeLenses)
+import Data.List.Index (indexed)
+import Safe (headMay)
+
+-- 'WalletData' like data type, used only in UI glue
+data UiWalletData = UiWalletData
+  { _uwdName     :: !Text
+  , _uwdId       :: !HdRootId
+  , _uwdAccounts :: !(Vector UiAccountData)
+  } deriving (Show, Generic)
+
+data UiAccountData = UiAccountData
+  { _uadName      :: !Text
+  , _uadPath      :: !Word32
+  -- Because of ChainType layer Now it should be ((HdAddressChain, Word32), Address) I guess, but
+  -- AFAIU we don't want a new layer, so addresses of both types wiil be in one list -- External first.
+  , _uadAddresses :: !(Vector (Word32, Address))
+  } deriving (Eq, Show, Generic)
+
+makeLenses 'UiWalletData
+makeLenses 'UiAccountData
+
+indexed_ :: (Integral i) => [a] -> [(i, a)]
+indexed_ x = mapFst fromIntegral <$> (indexed x)
```

Comment:
`enumerate` in serokell-util: https://hackage.haskell.org/package/serokell-util-0.8.0/docs/Serokell-Util-Common.html#v:enumerate

Reply:
Also, redundant parentheses around `Integral i`.